### PR TITLE
Add local view dropping/recreating to hotbackup.

### PR DESCRIPTION
**Pull request by neunhoef**

### CODE REVIEW

File: arangod/IResearch/IResearchFeature.cpp [hunk 1]
```diff
@@ -538,6 +539,179 @@ void registerTransactionDataSourceRegistrationCallback() {
...
+               "all ArangoSearch data for restore operation, msg: "
+            << res.errorMessage();
+        continue;
+      }
+    }
+
+    bool exists;
+
+    // remove any stale data-store
+    if (!dataPath.exists(exists) || (exists && !dataPath.remove())) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.cpp]: do not need to deal with index directory explicitly (see comment above)

File: arangod/IResearch/IResearchFeature.cpp [hunk 2]
```diff
@@ -538,6 +539,179 @@ void registerTransactionDataSourceRegistrationCallback() {
...
+    if (!res.ok()) {
+      LOG_TOPIC("12123", ERR, arangodb::iresearch::TOPIC)
+          << "failure to generate persisted definition while recreating "
+             "ArangoSearch index after a restore";
+
+      success = false;
+      continue;
+    }
+
+    irs::utf8_path dataPath;
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.cpp]: do not need to deal with index directory explicitly (see comment above)

File: arangod/IResearch/IResearchFeature.cpp [hunk 3]
```diff
@@ -538,6 +539,179 @@ void registerTransactionDataSourceRegistrationCallback() {
...
+             "after a restore";
+
+      success = false;
+      continue;
+    }
+
+    bool exists;
+
+    // remove any stale data-store
+    if (!dataPath.exists(exists) || (exists && !dataPath.remove())) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.cpp]: do not need to deal with index directory explicitly (see comment above)

File: arangod/IResearch/IResearchFeature.cpp [hunk 4]
```diff
@@ -968,6 +1142,14 @@ void IResearchFeature::validateOptions(std::shared_ptr<arangodb::options::Progra
   ApplicationFeature::validateOptions(options);
 }
 
+void IResearchFeature::removeLocalArangoSearchData() {
+  DatabaseFeature::DATABASE->enumerateDatabases(::removeAllArangoSearchDataForDatabase);
+}
+
+bool IResearchFeature::recreateLocalArangoSearchData(TRI_vocbase_t& vocbase) {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.cpp]: why not a static member/separate function?

File: arangod/IResearch/IResearchFeature.h [hunk 5]
```diff
@@ -77,6 +77,13 @@ class IResearchFeature final : public application_features::ApplicationFeature {
   void unprepare() override;
   void validateOptions(std::shared_ptr<options::ProgramOptions>) override;
 
+  //////////////////////////////////////////////////////////////////////////////
+  /// @brief remove local data, this is for a restore operation for hotbackup
+  //////////////////////////////////////////////////////////////////////////////
+
+  virtual void removeLocalArangoSearchData();
+  bool recreateLocalArangoSearchData(TRI_vocbase_t& vocbase);
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.h]: why not a static member/separate function?

File: arangod/RocksDBEngine/RocksDBHotBackup.cpp [hunk 6]
```diff
@@ -997,6 +998,22 @@ void RocksDBHotBackupRestore::execute() {
 
       restartAction = new std::function<int()>();
       *restartAction = localRestoreAction;
+
+      // Now remove all local ArangoSearch view data, since it will not be
+      // valid after the restore. Note that on a single server there is no
+      // automatism to recreate the data and on a dbserver, the Maintenance
+      // is stopped before we get here.
+      iresearch::IResearchFeature* arangoSearchFeature =
```
Comment (gnusi) [inline arangod/RocksDBEngine/RocksDBHotBackup.cpp]: honestly I don't understand why do we need feature instance for this. Local helper function which drops/recreates view is more than enough.

File: arangod/StorageEngine/HotBackupFeature.cpp [hunk 7]
```diff
@@ -23,23 +23,63 @@
...
+void recreateArangoSearchViewsAfterRestore() {
+  auto* arangoSearchFeature =
+        arangodb::application_features::ApplicationServer::lookupFeature<arangodb::iresearch::IResearchFeature>("ArangoSearch");
+  LOG_TOPIC("fdeda", INFO, Logger::BACKUP)
+    << "Recreating ArangoSearch indexes...";
+  DatabaseFeature::DATABASE->enumerateDatabases(
+    [arangoSearchFeature](TRI_vocbase_t& vocbase) {
+      LOG_TOPIC("efdab", INFO, Logger::BACKUP)
+        << "Recreating ArangoSearch index for database " << vocbase.name();
+      bool res = arangoSearchFeature->recreateLocalArangoSearchData(vocbase);
```
Comment (gnusi) [inline arangod/StorageEngine/HotBackupFeature.cpp]: honestly I don't understand why do we need feature instance for this. Local helper function which drops/recreates view is more than enough.

File: arangod/StorageEngine/HotBackupFeature.cpp [hunk 8]
```diff
@@ -330,5 +485,9 @@ bool HotBackupFeature::cancelled(std::string const& transferId) const {
   return false;
 }
 
+void HotBackupFeature::removeAllArangoSearchData() {
+  ::removeAllArangoSearchData();
```
Comment (gnusi) [inline arangod/StorageEngine/HotBackupFeature.cpp]: it seems having delegating method is redundant, why don't to implement it here

File: arangod/IResearch/IResearchFeature.h [hunk 9]
```diff
@@ -43,7 +43,7 @@ class ResourceMutex;  // forward declaration
 bool isFilter(arangodb::aql::Function const& func) noexcept;
 bool isScorer(arangodb::aql::Function const& func) noexcept;
 
-class IResearchFeature final : public application_features::ApplicationFeature {
```
Comment (gnusi) [inline arangod/IResearch/IResearchFeature.h]: I think `IResearchFeature` can stay `final`

File: arangod/StorageEngine/HotBackupFeature.cpp (hunk 8 above)
Comment (neunhoef) [reply arangod/StorageEngine/HotBackupFeature.cpp]: Agreed, will do the change.

File: arangod/IResearch/IResearchFeature.h (hunk 9 above)
Comment (neunhoef) [reply arangod/IResearch/IResearchFeature.h]: Right, will change back.

**Pull request merged**